ring_theory/euclidean_domain.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2018 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Chris Hughes
+-/
+import algebra.euclidean_domain ring_theory.ideals ring_theory.associated
+noncomputable theory
+local attribute [instance] classical.dec
+open euclidean_domain set ideal
+
+theorem span_gcd {α} [euclidean_domain α] (x y : α) :
+  span ({gcd x y} : set α) = span ({x, y} : set α) :=
+begin
+  apply le_antisymm; refine span_le.1 _,
+  { simp [submodule.span_span, mem_span_pair, submodule.le_def', mem_span_singleton'],
+    assume a b ha,
+    exact ⟨b * gcd_a x y, b * gcd_b x y, by rw [← ha, gcd_eq_gcd_ab x y];
+      simp [mul_add, mul_comm, mul_left_comm]⟩ },
+  { assume z ,
+    simp [mem_span_singleton, euclidean_domain.gcd_dvd_left, mem_span_pair,
+      @eq_comm _ _ z] {contextual := tt},
+    assume a b h,
+    exact dvd_add (dvd_mul_of_dvd_right (gcd_dvd_left _ _) _)
+      (dvd_mul_of_dvd_right (gcd_dvd_right _ _) _) }
+end
+
+theorem gcd_is_unit_iff {α} [euclidean_domain α] {x y : α} :
+  is_unit (gcd x y) ↔ is_coprime x y :=
+by rw [← span_singleton_eq_top, span_gcd, is_coprime]
+
+theorem is_coprime_of_dvd {α} [euclidean_domain α] {x y : α}
```

Comment:
I had a bit of a discussion with @ChrisHughes24 about this lemma and the next. Technically these lemmas are true in greater generality... (I guess in gcd domains). However the type class hierarchy in Lean is not as easy as the one on wiki (see https://en.wikipedia.org/wiki/GCD_domain). Therefore it seems that (for now) this might be a natural place to prove these facts.